File: 002.js
var _ = require("underscore");

function sumEvenFibs(upto) {
    return _.chain(generateFibonacciArray(upto))
        .select(function(fib) {
            return fib % 2 === 0;
        })
        .reduce(function(memo, evenFib) {
            return memo + evenFib;
        }, 0)
        .value();
}

function generateFibonacciArray(upto) {
    var arr = [];
    for (var i = 1; fibonacci(i) <= upto; i++) {
        arr.push(fibonacci(i));
    }
    return arr;
}

function fibonacci(n) {
    if (n < 2) {
        return 1;
    } else {
        return fibonacci(n - 2) + fibonacci(n - 1);
    }
}

console.log(sumEvenFibs(4000000));

File: 002.rb
def sum_even_fibonacci_numbers(upto)
  generate_fibonacci_array(upto).select(&:even?).reduce(:+)
end

def generate_fibonacci_array(upto)
  arr = []
  i = 0
  while fibonacci(i) < upto
    arr.push fibonacci(i)
    i += 1
  end
  arr
end

def fibonacci(n)
  if n == 0 || n == 1
    1
  else
    fibonacci(n-2) + fibonacci(n-1)
  end
end

def even?(n)
  n % 2 == 0
end

puts sum_even_fibonacci_numbers(4000000)
